### PR TITLE
WIP: bytes4 type

### PR DESCRIPTION
**Pull request by skellet0r**
### What I did

Add `bytes4` type.

Fixes: #1575 

### How I did it

tldr: Looked at stack traces ... 

Added in the definition and primitive classes, modified a few functions such as `clamp_basetype`, `abi_type_of`, to make sure the resulting `IR` was as I expected it to be.

TODO:

- Allow bytes4 literals without breaking `Bytes[4]` (if that's even possible)
- Test cases
- Fix ERC721 token example
- Docs
- conversions in vyper/builtin_functions/convert.py
- MethodID builtin compatibility

### How to verify it

*Test cases to be added*

### Description for the changelog

- Add bytes4 base type

### Cute Animal Picture

Put a link to a cute animal picture inside the parenthesis-->

> I think this'll be the last base type I add manually ... pretty sure moving to some more robust way to port all valid solidity types would be better :) (but I wanted this and uint8 sooner than later)

### CODE REVIEW

**Comment by charles-cooper** [comment]
closing since bytes4 was implemented in #2705
